tests/tt_metal/tt_metal/unit_tests/multichip/ring_gather_kernels.cpp

```diff
@@ -0,0 +1,483 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <functional>
+#include <random>
+
+#include "device_fixture.hpp"
+#include "n300_device_fixture.hpp"
+#include "tt_metal/detail/tt_metal.hpp"
+#include "tt_metal/host_api.hpp"
+#include "tt_metal/impl/kernels/kernel.hpp"
+#include "tt_metal/test_utils/comparison.hpp"
+#include "tt_metal/test_utils/df/df.hpp"
+#include "tt_metal/test_utils/print_helpers.hpp"
+#include "tt_metal/test_utils/stimulus.hpp"
+
+using namespace tt;
+using namespace tt::test_utils;
+using namespace tt::test_utils::df;
+
+constexpr std::int32_t WORD_SIZE = 16;  // 16 bytes per eth send packet
+constexpr std::int32_t MAX_NUM_WORDS =
+    (eth_l1_mem::address_map::MAX_L1_LOADING_SIZE - eth_l1_mem::address_map::ERISC_L1_UNRESERVED_BASE) / WORD_SIZE;
+
+struct BankedConfig {
+    size_t num_pages = 1;
+    size_t size_bytes = 1 * 2 * 32 * 32;
+    size_t page_size_bytes = 2 * 32 * 32;
+    BufferType input_buffer_type = BufferType::L1;
+    BufferType output_buffer_type = BufferType::L1;
+    tt::DataFormat l1_data_format = tt::DataFormat::Float16_b;
+};
+
+std::vector<int> get_hamiltonian_cycle(vector<vector<int>>& adj, int N, int s = 0) {
+    std::vector<std::vector<int>> dp(N, std::vector<int>(1 << N, -1));
+
+    for (int i = 0; i < N; ++i) {
+        if (adj[s][i]) {
+            dp[i][(1 << i)] = i;
+        }
+    }
+
+    for (int i = 0; i < (1 << N); ++i) {
+        for (int j = 0; j < N; ++j) {
+            if (i & (1 << j)) {
+                for (int k = 0; k < N; ++k) {
+                    if (i & (1 << k) && adj[k][j] && j != k && dp[k][i ^ (1 << j)] != -1) {
+                        dp[j][i] = k;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    for (int i = 0; i < N; ++i) {
+        int m = (1 << N) - 1;
+
+        if (dp[i][m] != -1 && i == s) {
+            std::vector<int> path;
+            path.reserve(N + 1);
+            path.push_back(i);
+
+            for (int j = 0; j < N - 1; ++j) {
+                path.push_back(dp[*path.rbegin()][m]);
+                m ^= 1 << *(path.rbegin() + 1);
+            }
+            path.push_back(s);
+            return path;
+        }
+    }
+    return {};
+}
+
+std::vector<Device*> get_device_ring(std::vector<tt::tt_metal::Device*> devices) {
+    std::vector<std::vector<int>> adj(devices.size(), std::vector<int>(devices.size(), 0));
+    for (uint32_t i = 0; i < devices.size(); ++i) {
+        const auto& device = devices[i];
+        for (const auto& connected_device_id : device->get_ethernet_connected_chip_ids()) {
```

Comment:
FYI, I made some changes to `get_ethernet_connected_chip_ids` to return a map of connect chips with their connected eth cores too. That PR will go in later than this one though